File: internal/web/outreach.go
package web

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/slaguardia/scout/internal/outreach"
	"github.com/slaguardia/scout/internal/store"
)

// OutreachRunner runs the outreach draft pipeline for one created draft row,
// asynchronously: Draft returns immediately and the pipeline writes its
// progress/result to the draft row (the panel polls). The engine owns models,
// retries, and the no-send route; the web layer owns the queue discipline (one
// active draft per posting) and the inputs-present gate.
type OutreachRunner interface {
	Draft(draftID int64)
}

// handlePostingOutreach is the draft queue on one posting:
//
//	GET  /api/postings/{id}/outreach  -> {drafts: [...]} newest first
//	POST /api/postings/{id}/outreach  -> start a draft (202 + the new row)
//
// POST gates on the two required inputs being present — the scout-local email
// template and a discovered experience bundle (the honesty ground truth) — and
// on no other active draft for the posting; it fails 503 when no engine is wired.
func (s *Server) handlePostingOutreach(w http.ResponseWriter, r *http.Request, postingID string) {
	switch r.Method {
	case http.MethodGet:
		drafts, err := s.DB.ListOutreachDrafts(postingID)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"drafts": drafts})

	case http.MethodPost:
		if s.Outreach == nil {
			http.Error(w, "outreach pipeline not wired (no engine in this build)", http.StatusServiceUnavailable)
			return
		}
		// GATE: a draft needs an experience bundle (the honesty ground truth).
		// The email template always exists (the DB row or the compiled-in
		// default), so it is never a gate.
		if exp, err := s.DB.OutreachKnowledge("experience"); err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		} else if strings.TrimSpace(exp) == "" {
			// Cold cache: sync from the brain once, then re-check before blocking —
			// a first-ever draft shouldn't be gated just because nothing's discovered yet.
			s.ensureOutreachKnowledge(r.Context())
			if exp2, _ := s.DB.OutreachKnowledge("experience"); strings.TrimSpace(exp2) == "" {
				writeJSON(w, http.StatusPreconditionFailed, map[string]any{
					"error": "no experience page found in your brain — add one; scout syncs it automatically",
					"need":  "experience",
				})
				return
			}
		}
		// Voice is soft: drafting proceeds without it (a less-voiced email).
		degraded := []string{}
		if v, _ := s.DB.OutreachKnowledge("voice"); strings.TrimSpace(v) == "" {
			degraded = append(degraded, "voice")
		}
		// ?regenerate=1 retires the current awaiting_review/needs_work/no_hook
		// draft and starts a fresh run (re-draft after backfilling info); the
		// default POST creates only when no draft is active (409 otherwise).
		create := s.DB.CreateOutreachDraft
		if r.URL.Query().Get("regenerate") == "1" {
			create = s.DB.RegenerateOutreachDraft
		}
		d, err := create(postingID)
		if err != nil {
			switch {
			case errors.Is(err, sql.ErrNoRows):
				http.NotFound(w, r)
			case strings.Contains(err.Error(), "active draft"):
				http.Error(w, err.Error(), http.StatusConflict)
			default:
				http.Error(w, err.Error(), http.StatusInternalServerError)
			}
			return
		}
		s.Outreach.Draft(d.ID)
		writeJSON(w, http.StatusAccepted, map[string]any{"draft": d, "degraded": degraded})

	default:
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
	}
}

// handleOutreach routes /api/outreach/*:
//
//	GET  /api/outreach/sources            -> the discovered knowledge sources (read-only)
//	PUT  /api/outreach/drafts/{id}        -> save the user's edit
//	POST /api/outreach/drafts/{id}/sent   -> mark sent (bumps posting tracking)
func (s *Server) handleOutreach(w http.ResponseWriter, r *http.Request) {
	rest := strings.TrimPrefix(r.URL.Path, "/api/outreach/")
	switch {
	case rest == "sources" || strings.HasPrefix(rest, "sources/"):
		s.handleOutreachSources(w, r, strings.TrimPrefix(strings.TrimPrefix(rest, "sources"), "/"))
	case strings.HasPrefix(rest, "drafts/"):
		s.handleDraft(w, r, strings.TrimPrefix(rest, "drafts/"))
	default:
		http.NotFound(w, r)
	}
}

// handleOutreachSources serves the discovered knowledge sources, read-only — a
// diagnostic peek at what the brain resolved for each need. There is no refresh
// or add/remove: the bundle syncs from the brain automatically (see
// ensureOutreachKnowledge), so the dashboard never edits it.
func (s *Server) handleOutreachSources(w http.ResponseWriter, r *http.Request, action string) {
	if action != "" || r.Method != http.MethodGet {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}
	writeJSON(w, http.StatusOK, s.sourcesPayload())
}

// ensureOutreachKnowledge runs the change-aware brain sync for the outreach
// knowledge bundle (best-effort), resolving a dashboard-stored API key first so
// discovery can run. A no-op when the brain isn't configured; any failure leaves
// the last-good cache in place (the caller re-checks the cache before gating).
func (s *Server) ensureOutreachKnowledge(ctx context.Context) {
	if s.Brainbot == nil || !s.Brainbot.Enabled() {
		return
	}
	s.ensureAnthropicKey() // DB-over-env, so a dashboard-stored key lights up discovery
	ctx, cancel := context.WithTimeout(ctx, 90*time.Second)
	defer cancel()
	_ = outreach.EnsureKnowledge(ctx, s.Brainbot, s.Anthropic, s.DB, "", nil)
}

// sourcesPayload lists the cached sources without their (large) content — the
// per-need pointers the UI renders.
func (s *Server) sourcesPayload() map[string]any {
	srcs, err := s.DB.ListOutreachSources()
	if err != nil {
		return map[string]any{"sources": []any{}, "error": err.Error()}
	}
	type lite struct {
		Need       string `json:"need"`
		PageID     string `json:"page_id"`
		Title      string `json:"title"`
		Version    string `json:"version"`
		ResolvedAt string `json:"resolved_at"`
	}
	out := make([]lite, 0, len(srcs))
	for _, s := range srcs {
		out = append(out, lite{s.Need, s.PageID, s.Title, s.Version, s.ResolvedAt})
	}
	return map[string]any{"sources": out, "needs": outreach.KnowledgeNeeds}
}

// lintBody drops a leading "Subject:" line before the voice flag, since the
// subject's em dash ("Name — intro") is intentional and not a violation.
func lintBody(email string) string {
	if strings.HasPrefix(email, "Subject:") {
		if i := strings.IndexByte(email, '\n'); i >= 0 {
			return email[i+1:]
		}
	}
	return email
}

func (s *Server) handleDraft(w http.ResponseWriter, r *http.Request, rest string) {
	idStr, action, _ := strings.Cut(strings.Trim(rest, "/"), "/")
	id, err := strconv.ParseInt(idStr, 10, 64)
	if err != nil {
		http.NotFound(w, r)
		return
	}

	switch {
	case action == "" && r.Method == http.MethodGet:
		d, err := s.DB.GetOutreachDraft(id)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		if d == nil {
			http.NotFound(w, r)
			return
		}
		writeJSON(w, http.StatusOK, d)

	case action == "" && r.Method == http.MethodPut:
		var body struct {
			Edited string `json:"edited"`
		}
		if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, 32<<10)).Decode(&body); err != nil {
			http.Error(w, "invalid JSON: "+err.Error(), http.StatusBadRequest)
			return
		}
		// Edits are only meaningful pre-send: a sent draft is the record of what
		// was actually emailed, and a researching one is pipeline-owned. A
		// needs_work draft is finished and reviewable — just flagged by the judge.
		if cur, err := s.DB.GetOutreachDraft(id); err == nil && cur != nil &&
			cur.Status != store.DraftAwaitingReview && cur.Status != store.DraftNeedsWork && cur.Status != store.DraftNoHook {
			http.Error(w, "draft is "+cur.Status+" — only awaiting_review/needs_work/no_hook drafts are editable", http.StatusConflict)
			return
		}
		// Re-run the deterministic voice flag on the edit (body only — the subject
		// line's em dash is intentional), plus the word-count check. Non-blocking:
		// it just refreshes the chips.
		findings := outreach.VoiceFindings(lintBody(body.Edited))
		findings = append(findings, outreach.LengthFindings(body.Edited)...)
		if findings == nil {
			findings = []outreach.LintFinding{}
		}
		lintJSON, _ := json.Marshal(findings)
		if err := s.DB.SetOutreachDraftEdited(id, body.Edited, string(lintJSON)); err != nil {
			if errors.Is(err, sql.ErrNoRows) {
				http.NotFound(w, r)
				return
			}
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		d, err := s.DB.GetOutreachDraft(id)
		if err != nil || d == nil {
			http.Error(w, "draft vanished", http.StatusInternalServerError)
			return
		}
		writeJSON(w, http.StatusOK, d)

	case action == "sent" && r.Method == http.MethodPost:
		d, err := s.DB.MarkOutreachDraftSent(id)
		if err != nil {
			if errors.Is(err, sql.ErrNoRows) {
				http.NotFound(w, r)
				return
			}
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		writeJSON(w, http.StatusOK, d)

	default:
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
	}
}

File: scout/web/routes/outreach.py
"""The outreach draft queue + the discovered-sources peek + draft edit/send.

The async pipeline is fire-and-forget:
the POST creates a draft row and calls state.outreach.draft(id, skip) — the runner
(the wired Engine, or a test fake) owns the background thread and its OWN
connection; the request connection here is only used for the gate + the draft-row
create/list. The panel polls the draft row for progress.
"""

from __future__ import annotations

import json

from fastapi import APIRouter, Depends, Request
from starlette.responses import Response

from scout import outreach as outreach_pkg
from scout.store import contacts, errors, outreach_drafts, outreach_sources

from ..deps import AppState, get_db, get_state
from ..responses import json_error, json_response
from .core import _s, decode_json, raw_body

router = APIRouter()


# --- knowledge sync ----------------------------------------------------------


def _ensure_outreach_knowledge(con, state: AppState) -> None:
    """Best-effort change-aware brain sync for the outreach knowledge bundle. A
    no-op when the brain isn't configured; any failure leaves the last-good cache
    in place (the caller re-checks the cache before gating)."""
    if state.brainbot is None or not state.brainbot.enabled():
        return
    state.ensure_anthropic_key(con)  # DB-over-env, so a dashboard key lights up discovery
    try:
        outreach_pkg.ensure_knowledge(state.brainbot, state.anthropic, con, "", None)
    except Exception:  # noqa: BLE001 - best-effort; serve the cache
        pass


def _experience_gate(con, state: AppState) -> Response | None:
    """The shared draft/answer honesty gate: an empty experience bundle triggers a
    one-shot brain sync, then (still empty) a 412 with need=experience. Returns the
    412 Response to send, or None when the gate passes."""
    exp = outreach_sources.outreach_knowledge(con, "experience")
    if exp.strip() != "":
        return None
    # Cold cache: sync from the brain once, then re-check before blocking.
    _ensure_outreach_knowledge(con, state)
    if outreach_sources.outreach_knowledge(con, "experience").strip() != "":
        return None
    return json_response(
        {
            "error": "no experience page found in your brain — add one; scout syncs it automatically",
            "need": "experience",
        },
        412,
    )


# --- the draft queue on one posting: /api/postings/{id}/outreach -------------


@router.get("/api/postings/{posting_id}/outreach")
def list_posting_outreach(posting_id: str, con=Depends(get_db)) -> Response:
    """The posting's drafts, newest first."""
    return json_response({"drafts": outreach_drafts.list_outreach_drafts(con, posting_id)})


@router.post("/api/postings/{posting_id}/outreach")
def start_posting_outreach(
    posting_id: str,
    request: Request,
    con=Depends(get_db),
    state: AppState = Depends(get_state),
) -> Response:
    """Start (or regenerate) a draft. Gates on the experience bundle (the honesty
    ground truth) and one active draft per posting; 503 when no engine is wired."""
    if state.outreach is None:
        return json_error("outreach pipeline not wired (no engine in this build)", 503)

    gate = _experience_gate(con, state)
    if gate is not None:
        return gate

    # Voice is soft: drafting proceeds without it (a less-voiced email).
    degraded: list[str] = []
    if outreach_sources.outreach_knowledge(con, "voice").strip() == "":
        degraded.append("voice")

    # ?research=0 skips the web-research stage for this draft (drafts from on-file
    # info only). Persisted on the row so the panel's progress bar can drop the
    # Research node across polls/reloads.
    skip_research = request.query_params.get("research") == "0"

    # ?regenerate=1 retires the current reviewable draft and starts a fresh run;
    # the default POST creates only when no draft is active (409 otherwise).
    create = outreach_drafts.create_outreach_draft
    if request.query_params.get("regenerate") == "1":
        create = outreach_drafts.regenerate_outreach_draft
    try:
        d = create(con, posting_id, skip_research)
    except errors.NotFound:
        return json_error("not found", 404)
    except ValueError as e:
        if "active draft" in str(e):
            return json_error(str(e), 409)
        raise

    state.outreach.draft(d.id, skip_research)
    return json_response({"draft": d, "degraded": degraded}, 202)


# --- discovered sources (read-only): /api/outreach/sources -------------------


@router.get("/api/outreach/sources")
def outreach_sources_endpoint(con=Depends(get_db)) -> Response:
    """The cached knowledge sources without their (large) content — the per-need
    pointers the UI renders. Read-only: the bundle auto-syncs from the brain."""
    srcs = outreach_sources.list_outreach_sources(con)
    lite = [
        {
            "need": s.need,
            "page_id": s.page_id,
            "title": s.title,
            "version": s.version,
            "resolved_at": s.resolved_at,
        }
        for s in srcs
    ]
    return json_response({"sources": lite, "needs": outreach_pkg.KNOWLEDGE_NEEDS})


# --- one draft: /api/outreach/drafts/{id}[/sent] -----------------------------


def _parse_int_id(raw_id: str) -> int | None:
    try:
        return int(raw_id)
    except (TypeError, ValueError):
        return None


def _lint_body(email: str) -> str:
    """Drop a leading "Subject:" line before the voice flag, since the subject's
    em dash ("Name — intro") is intentional and not a violation."""
    if email.startswith("Subject:"):
        i = email.find("\n")
        if i >= 0:
            return email[i + 1 :]
    return email


@router.get("/api/outreach/drafts/{raw_id}")
def get_draft(raw_id: str, con=Depends(get_db)) -> Response:
    id = _parse_int_id(raw_id)
    if id is None:
        return json_error("not found", 404)
    d = outreach_drafts.get_outreach_draft(con, id)
    if d is None:
        return json_error("not found", 404)
    return json_response(d)


@router.put("/api/outreach/drafts/{raw_id}")
def save_draft_edit(raw_id: str, raw: bytes = Depends(raw_body), con=Depends(get_db)) -> Response:
    """Save the user's edit. Only awaiting_review/needs_work/no_hook drafts are
    editable (a sent draft is the record of what was emailed; a researching one is
    pipeline-owned) → 409 otherwise."""
    id = _parse_int_id(raw_id)
    if id is None:
        return json_error("not found", 404)
    body = decode_json(raw)
    edited = _s(body, "edited")

    cur = outreach_drafts.get_outreach_draft(con, id)
    editable = (
        outreach_drafts.DRAFT_AWAITING_REVIEW,
        outreach_drafts.DRAFT_NEEDS_WORK,
        outreach_drafts.DRAFT_NO_HOOK,
    )
    if cur is not None and cur.status not in editable:
        return json_error(
            f"draft is {cur.status} — only awaiting_review/needs_work/no_hook drafts are editable",
            409,
        )

    # Re-run the deterministic voice flag (body only) + the word-count check.
    findings = outreach_pkg.voice_findings(_lint_body(edited)) + outreach_pkg.length_findings(
        edited
    )
    lint_json = json.dumps(
        [{"code": f.code, "message": f.message} for f in findings], separators=(",", ":")
    )
    outreach_drafts.set_outreach_draft_edited(con, id, edited, lint_json)
    d = outreach_drafts.get_outreach_draft(con, id)
    if d is None:
        return json_error("draft vanished", 500)
    return json_response(d)


@router.post("/api/outreach/drafts/{raw_id}/sent")
def mark_draft_sent(raw_id: str, raw: bytes = Depends(raw_body), con=Depends(get_db)) -> Response:
    """Mark a draft sent (idempotent). 404 unknown.

    With a {"contact_id": ...} body, also log the send against that contact —
    arming its follow-up and seeding the posting's outreach_status — so a send
    you made by hand is tracked exactly like a Gmail send, minus the live thread
    link. Without a contact_id it stays a bare status flip (back-compat).

    An already-sent draft can be reused to reach out to *another* contact: it logs
    the new contact, leaving the terminal status intact. A contact already on this
    posting's log is skipped (idempotent re-mark; no duplicate send)."""
    id = _parse_int_id(raw_id)
    if id is None:
        return json_error("not found", 404)
    body = decode_json(raw) if raw.strip() else {}
    contact_id = _s(body, "contact_id").strip()
    if contact_id:
        d = outreach_drafts.get_outreach_draft(con, id)
        if d is None:
            return json_error("not found", 404)
        # Log the send unless this contact was already logged for the posting —
        # the contact-level guard (replacing the old "already sent → skip") is what
        # lets a sent draft be reused for someone else while staying idempotent.
        prior = contacts.list_outreach_for_posting(con, d.posting_id)
        if not any(e.contact_id == contact_id for e in prior):
            contact = contacts.get_contact(con, contact_id)
            recipient = (contact.name or "").strip().split(" ")[0] if contact else ""
            draft_text = d.edited if d.edited.strip() else d.draft
            if recipient:
                draft_text = draft_text.replace("[Recipient]", recipient)
            try:
                contacts.log_outreach(
                    con, d.posting_id, contact_id, contacts.OutreachInput(body=draft_text)
                )
            except ValueError as e:
                return json_error(str(e), 400)
    out = outreach_drafts.mark_outreach_draft_sent(con, id)
    return json_response(out)


@router.post("/api/outreach/drafts/{raw_id}/cancel")
def cancel_draft(raw_id: str, con=Depends(get_db)) -> Response:
    """Cancel a running (researching) draft — delete its row so the posting's
    active-draft slot frees up and the panel can start over. 200 with
    {"cancelled": bool}; false when the draft already finished (nothing to cancel).
    The background pipeline thread finishes on its own; its writes no-op once the
    row is gone."""
    id = _parse_int_id(raw_id)
    if id is None:
        return json_error("not found", 404)
    cancelled = outreach_drafts.cancel_outreach_draft(con, id)
    return json_response({"cancelled": cancelled})


@router.delete("/api/outreach/drafts/{raw_id}")
def delete_draft(raw_id: str, con=Depends(get_db)) -> Response:
    """Delete a draft from the queue/history (any status). 200 with
    {"deleted": bool}; false when the id is already gone. A researching draft is
    deleted too — same as cancel — so the background thread's writes then no-op."""
    id = _parse_int_id(raw_id)
    if id is None:
        return json_error("not found", 404)
    deleted = outreach_drafts.delete_outreach_draft(con, id)
    return json_response({"deleted": deleted})
